mirrord/layer/src/socket/hooks.rs

```diff
@@ -370,6 +371,128 @@ pub(super) unsafe extern "C" fn send_to_detour(
         })
     }
 }
+
+#[hook_guard_fn]
+pub(super) unsafe extern "C" fn recvmsg_detour(
+    sockfd: i32,
+    message_header: *mut libc::msghdr,
+    flags: c_int,
+) -> ssize_t {
+    // Returns `-1`, but calling the original function sets whatever `errno` for us.
+    if message_header.is_null() {
+        libc::recvmsg(sockfd, message_header, flags)
+    } else {
+        if (*message_header).msg_name.is_null() {
+            // Safety: we can deref `message_header` here, as we check for null before.
+            let data_blocks = (*message_header).msg_iov;
+            let number_of_blocks = (*message_header).msg_iovlen;
+
+            let mut offset = 0;
+            let total_length = (0..number_of_blocks)
+                .reduce(|acc, _| {
+                    let block = data_blocks.byte_add(offset);
+                    offset += (*data_blocks).iov_len;
+
+                    acc + (*block).iov_len
+                })
+                .unwrap_or_default();
+            // TODO(alex) [mid] 2023-08-08: Do we need this `total_length` check though? Real use
+            // example will tell us the answer.
+            let layout = Layout::array::<u8>(total_length).unwrap();
```

Comment:
You can use more high level API for interacting with the cmsg.
https://docs.rs/nix/0.26.2/nix/sys/socket/enum.ControlMessageOwned.html

Reply:
Nice, had no idea about this.

Reply:
For now, it seems like we're safe to ignore this stuff.